src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -124,20 +126,54 @@ void CALLBACK cfApiFetchDataCallback(const CF_CALLBACK_INFO *callbackInfo, const
     }
 
     qint64 offset = 0;
+    QByteArray protrudingData;
 
     QObject::connect(&socket, &QLocalSocket::readyRead, &loop, [&] {
-        auto data = socket.readAll();
-        if (data.isEmpty()) {
+        const auto dataReceived = socket.readAll();
+        if (dataReceived.isEmpty()) {
             qCWarning(lcCfApiWrapper) << "Unexpected empty data received" << requestId;
             sendTransferError();
             loop.quit();
             return;
         }
-        sendTransferInfo(data, offset);
-        offset += data.length();
+
+        // CFAPI expects sent blocks to be of a multiple of a block size.
+        // Only the last sent block is allowed to be of a different size than
+        // a multiple of a block size
+        constexpr auto cfapiBlockSize = 4096;
+
+        const QByteArray data = protrudingData + dataReceived;
+        protrudingData.clear();
+
+        QByteArray dataToSend;
+        // Extract blocks from data if possible
+        const auto blocksToSendCount = (data.size() - 1) / cfapiBlockSize;
```

Comment:
@FlexW Pay attention - if you divide data.size() - 1 to 4096 - you will end up having one block less than it is available in the data. This is because of this -1. Let's say data.size() is 4096. Then, it effectively becomes 4096 after subtraction. Then, 4095/4096 would give you 0.

Reply:
Of course you are correct. Thank you. That mistake comes from refactoring. Sadly the overall logic is still correct I think, so this mistake could not be observed while testing.

Reply:
> Of course you are correct. Thank you. That mistake comes from refactoring. Sadly the overall logic is still correct I think, so this mistake could not be observed while testing.

@FlexW This may have worked because you have the omnipresent **protrudingData** that fixes it all!